### Import Libraries

In [1]:
import numpy as np
import pandas as pd

In [2]:
df = pd.read_csv('fifa21 data.csv')
pd.set_option('display.max_columns', None)
df

C:\Users\ASUS\AppData\Local\Temp\ipykernel_11708\887213838.py:1: DtypeWarning: Columns (76) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('fifa21 data.csv')


ID               Name                      LongName  \
0      158023           L. Messi                  Lionel Messi   
1       20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2      200389           J. Oblak                     Jan Oblak   
3      192985       K. De Bruyne               Kevin De Bruyne   
4      190871          Neymar Jr    Neymar da Silva Santos Jr.   
...       ...                ...                           ...   
18974  247223             Xia Ao                        Ao Xia   
18975  258760           B. Hough                     Ben Hough   
18976  252757        R. McKinley                Ronan McKinley   
18977  243790       Wang Zhen'ao                  Zhen'ao Wang   
18978  252520          Zhou Xiao                     Xiao Zhou   

                                               photoUrl  \
0      https://cdn.sofifa.com/players/158/023/21_60.png   
1      https://cdn.sofifa.com/players/020/801/21_60.png   
2      https://cdn.sofifa.com/players/200/389/21_60.png   
3      https://cdn.sofifa.com/players/192/985/21_60.png   
4      https://cdn.sofifa.com/players/190/871/21_60.png   
...                                                 ...   
18974  https://cdn.sofifa.com/players/247/223/21_60.png   
18975  https://cdn.sofifa.com/players/258/760/21_60.png   
18976  https://cdn.sofifa.com/players/252/757/21_60.png   
18977  https://cdn.sofifa.com/players/243/790/21_60.png   
18978  https://cdn.sofifa.com/players/252/520/21_60.png   

                                               playerUrl Nationality  Age  \
0      http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33   
1      http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35   
2      http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27   
3      http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29   
4      http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28   
...                                                  ...         ...  ...   
18974     http://sofifa.com/player/247223/ao-xia/210006/    China PR   21   
18975  http://sofifa.com/player/258760/ben-hough/210006/     England   17   
18976  http://sofifa.com/player/252757/ronan-mckinley...     England   18   
18977  http://sofifa.com/player/243790/zhenao-wang/21...    China PR   20   
18978  http://sofifa.com/player/252520/xiao-zhou/210006/    China PR   21   

       ↓OVA  POT                         Club     Contract   Positions Height  \
0        93   93         \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   
1        92   92             \n\n\n\nJuventus  2018 ~ 2022      ST, LW  187cm   
2        91   93      \n\n\n\nAtlético Madrid  2014 ~ 2023          GK  188cm   
3        91   91      \n\n\n\nManchester City  2015 ~ 2023     CAM, CM  181cm   
4        91   91  \n\n\n\nParis Saint-Germain  2017 ~ 2022     LW, CAM  175cm   
...     ...  ...                          ...          ...         ...    ...   
18974    47   55           \n\n\n\nWuhan Zall  2018 ~ 2022          CB  178cm   
18975    47   67      \n\n\n\nOldham Athletic  2020 ~ 2021          CM  175cm   
18976    47   65           \n\n\n\nDerry City  2019 ~ 2020          CM  179cm   
18977    47   57     \n\n\n\nDalian YiFang FC  2020 ~ 2022          RW  175cm   
18978    47   57     \n\n\n\nDalian YiFang FC  2019 ~ 2023      CB, LB  188cm   

      Weight Preferred Foot  BOV Best Position     Joined Loan Date End  \
0       72kg           Left   93            RW  01-Jul-04           NaN   
1       83kg          Right   92            ST  10-Jul-18           NaN   
2       87kg          Right   91            GK  16-Jul-14           NaN   
3       70kg          Right   91           CAM  30-Aug-15           NaN   
4       68kg          Right   91            LW  03-Aug-17           NaN   
...      ...            ...  ...           ...        ...           ...   
18974   66kg          Right   49            CB  13-Jul-18           NaN   
18975   65k

In [3]:
df.shape

(18979, 77)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 77 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                18979 non-null  int64 
 1   Name              18979 non-null  object
 2   LongName          18979 non-null  object
 3   photoUrl          18979 non-null  object
 4   playerUrl         18979 non-null  object
 5   Nationality       18979 non-null  object
 6   Age               18979 non-null  int64 
 7   ↓OVA              18979 non-null  int64 
 8   POT               18979 non-null  int64 
 9   Club              18979 non-null  object
 10  Contract          18979 non-null  object
 11  Positions         18979 non-null  object
 12  Height            18979 non-null  object
 13  Weight            18979 non-null  object
 14  Preferred Foot    18979 non-null  object
 15  BOV               18979 non-null  int64 
 16  Best Position     18979 non-null  object
 17  Joined      

# Data Cleaning

### Create a copy of DataFrame

In [5]:
data = df.copy()
data.head(5)

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389           J. Oblak                     Jan Oblak   
3  192985       K. De Bruyne               Kevin De Bruyne   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   
4  https://cdn.sofifa.com/players/190/871/21_60.png   

                                           playerUrl Nationality  Age  ↓OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27    91   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29    91   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28    91   

   POT                         Club     Contract   Positions Height Weight  \
0   93         \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   72kg   
1   92             \n\n\n\nJuventus  2018 ~ 2022      ST, LW  187cm   83kg   
2   93      \n\n\n\nAtlético Madrid  2014 ~ 2023          GK  188cm   87kg   
3   91      \n\n\n\nManchester City  2015 ~ 2023     CAM, CM  181cm   70kg   
4   91  \n\n\n\nParis Saint-Germain  2017 ~ 2022     LW, CAM  175cm   68kg   

  Preferred Foot  BOV Best Position     Joined Loan Date End    Value   Wage  \
0           Left   93            RW  01-Jul-04           NaN  €103.5M  €560K   
1          Right   92            ST  10-Jul-18           NaN     €63M  €220K   
2          Right   91            GK  16-Jul-14           NaN    €120M  €125K   
3          Right   91           CAM  30-Aug-15           NaN    €129M  €370K   
4          Right   91            LW  03-Aug-17           NaN    €132M  €270K   

  Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0        €138.4M        429        85         95                70   
1         €75.9M        437        84         95                90   
2        €159.4M         95        13         11                15   
3          €161M        407        94         82                55   
4        €166.5M        408        85         87                62   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   
2             43       13    109         12     13           14            40   
3             94       82    441         88     85           83            93   
4             87       87    448         95     88           89            81   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   
2            30       307            43            60       67         88   
3            92       398            77            76       78         91   
4            95       453            94            89       96         91   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   
2       49    268          59       78       41        78          12   
3       76    408          91       63       89        74          91   
4       83    357          80       62       81        50          84   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalt

### 1. Club

Current for each player.

In [6]:
data['Club'].dtype

dtype('O')

In [7]:
data['Club'].unique()

array(['\n\n\n\nFC Barcelona', '\n\n\n\nJuventus',
       '\n\n\n\nAtlético Madrid', '\n\n\n\nManchester City',
       '\n\n\n\nParis Saint-Germain', '\n\n\n\nFC Bayern München',
       '\n\n\n\nLiverpool', '\n\n\n\nReal Madrid', '\n\n\n\nChelsea',
       '\n\n\n\nTottenham Hotspur', '\n\n\n\nInter', '\n\n\n\nNapoli',
       '\n\n\n\nBorussia Dortmund', '\n\n\n\nManchester United',
       '\n\n\n\nArsenal', '\n\n\n\nLazio', '\n\n\n\nLeicester City',
       '\n\n\n\nBorussia Mönchengladbach', '\n\n\n\nReal Sociedad',
       '\n\n\n\nAtalanta', '\n\n\n\nOlympique Lyonnais', '\n\n\n\nMilan',
       '\n\n\n\nVillarreal CF', '\n\n\n\nRB Leipzig', '\n\n\n\nCagliari',
       '\n\n\n\nAjax', '\n\n\n\nSL Benfica', '\n\n\n\nAS Monaco',
       '\n\n\n\nWolverhampton Wanderers', '\n\n\n\nEverton',
       '\n\n\n\nFiorentina', '\n\n\n\nFC Porto', '\n\n\n\nRC Celta',
       '\n\n\n\nTorino', '\n\n\n\nSevilla FC', '\n\n\n\nGrêmio',
       '\n\n\n\nReal Betis', '\n\n\n\nRoma', '\n\n\n\nNewcastle Unite

There are many whitespace characters in some club names in the Club column. Since the whitespace characters are both at the beginning and end of the string, using .strip() would be appropriate to remove them.

In [8]:
data['Club'] = data['Club'].str.strip()
data['Club'].unique()

array(['FC Barcelona', 'Juventus', 'Atlético Madrid', 'Manchester City',
       'Paris Saint-Germain', 'FC Bayern München', 'Liverpool',
       'Real Madrid', 'Chelsea', 'Tottenham Hotspur', 'Inter', 'Napoli',
       'Borussia Dortmund', 'Manchester United', 'Arsenal', 'Lazio',
       'Leicester City', 'Borussia Mönchengladbach', 'Real Sociedad',
       'Atalanta', 'Olympique Lyonnais', 'Milan', 'Villarreal CF',
       'RB Leipzig', 'Cagliari', 'Ajax', 'SL Benfica', 'AS Monaco',
       'Wolverhampton Wanderers', 'Everton', 'Fiorentina', 'FC Porto',
       'RC Celta', 'Torino', 'Sevilla FC', 'Grêmio', 'Real Betis', 'Roma',
       'Newcastle United', 'Eintracht Frankfurt', 'Valencia CF',
       'Medipol Başakşehir FK', 'Inter Miami', 'Bayer 04 Leverkusen',
       'Levante UD', 'Crystal Palace', 'Athletic Club de Bilbao',
       'Shanghai SIPG FC', 'VfL Wolfsburg',
       'Guangzhou Evergrande Taobao FC', 'Al Shabab',
       'Olympique de Marseille', 'Los Angeles FC',
       'Beijing Sino

### 2. Contract

In [9]:
data['Contract'].dtype

dtype('O')

In [10]:
data['Contract'].unique()

array(['2004 ~ 2021', '2018 ~ 2022', '2014 ~ 2023', '2015 ~ 2023',
       '2017 ~ 2022', '2017 ~ 2023', '2018 ~ 2024', '2014 ~ 2022',
       '2018 ~ 2023', '2016 ~ 2023', '2013 ~ 2023', '2011 ~ 2023',
       '2009 ~ 2022', '2005 ~ 2021', '2011 ~ 2021', '2015 ~ 2022',
       '2017 ~ 2024', '2010 ~ 2024', '2012 ~ 2021', '2019 ~ 2024',
       '2015 ~ 2024', '2017 ~ 2025', '2020 ~ 2025', '2019 ~ 2023',
       '2008 ~ 2023', '2015 ~ 2021', '2020 ~ 2022', '2012 ~ 2022',
       '2016 ~ 2025', '2013 ~ 2022', '2011 ~ 2022', '2012 ~ 2024',
       '2016 ~ 2021', '2012 ~ 2023', '2008 ~ 2022', '2019 ~ 2022',
       '2017 ~ 2021', '2013 ~ 2024', '2020 ~ 2024', '2010 ~ 2022',
       '2020 ~ 2021', '2011 ~ 2024', '2020 ~ 2023', '2014 ~ 2024',
       '2013 ~ 2026', '2016 ~ 2022', '2010 ~ 2021', '2013 ~ 2021',
       '2019 ~ 2025', '2018 ~ 2025', '2016 ~ 2024', '2018 ~ 2021',
       '2009 ~ 2024', '2007 ~ 2022', 'Jun 30, 2021 On Loan',
       '2009 ~ 2021', '2019 ~ 2021', '2019 ~ 2026', 'Free', '2012 ~ 

using `.loc` to search for values in the 'Contract' column for each row where the value contains the string 'On Loan' or 'Free'.

In [11]:
for index, kontrak in data.loc[(data['Contract'].str.contains('On Loan')) | (data['Contract'].str.contains('Free')), 'Contract'].items():
    print(kontrak)

Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Free
Free
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Free
Free
Free
Free
Free
Free
Free
Free
Free
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Free
Free
Free
Free
Jun 30, 2021 On Loan
Dec 31, 2020 On Loan
Jun 30, 2021 On Loan
Free
Jun 30, 2021 On Loan
Free
Free
Free
Free
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Free
Jun 30, 2021 On Loan
Free
Jun 30, 2021 On Loan
Jan 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Free
Dec 31, 2020 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Free
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Jun 30, 2021 On Loan
Free
Jun 30, 2021 On Loan
Free
Free
Free

Declare a function `extract_contract` that accepts one parameter `contract`, which is used to create 3 new columns: 'Contract Start', 'Contract End', and 'Contract Length (years)' with values taken from the 'Contract' column.

In [12]:
def extract_contract(contract):
    if contract == 'Free' or 'On Loan' in contract:
        start_date = np.nan
        end_date = np.nan
        contract_length = 0
    else:
        start_date, end_date = contract.split(' ~ ')
        start_year = int(start_date[:4])
        end_year = int(end_date[:4])
        contract_length = end_year - start_year
    return start_date, end_date, contract_length

new_columns = ['Contract Start', 'Contract End', 'Contract Length(years)']
new_data = data['Contract'].apply(lambda x: pd.Series(extract_contract(x)))

for i in range(len(new_columns)):
    data.insert(loc=data.columns.get_loc('Contract')+1+i, column=new_columns[i], value=new_data[i])

In [13]:
data[['Contract', 'Contract Start', 'Contract End', 'Contract Length(years)']].sample(10)

Contract Contract Start Contract End  Contract Length(years)
18235  2017 ~ 2020           2017         2020                     3.0
12117  2019 ~ 2021           2019         2021                     2.0
2054   2018 ~ 2022           2018         2022                     4.0
6512   2020 ~ 2021           2020         2021                     1.0
10047  2018 ~ 2022           2018         2022                     4.0
12134  2020 ~ 2022           2020         2022                     2.0
4805          Free            NaN          NaN                     0.0
15276  2019 ~ 2023           2019         2023                     4.0
10600  2019 ~ 2021           2019         2021                     2.0
7011   2015 ~ 2020           2015         2020                     5.0

Declare a function `Categorize_Contract` that categorizes values in the 'Contract' column into three categories: 'Free', 'On Loan', and 'Contract'. Then create a new column 'Contract Status' to the DataFrame data to the right of the 'Contract Length (years)' column. The values for this new column are generated by applying the `Categorize_Contract` function to each value in the 'Contract' column.

In [14]:
def Categorize_Contract(contract):
    if contract == 'Free':
        return 'Free'
    elif 'On Loan'in contract:
        return 'On Loan'
    else:
        return 'Contract'

data.insert(data.columns.get_loc('Contract Length(years)')+1, 'Contract Status', data['Contract'].apply(Categorize_Contract))
data.sample(4)

ID          Name         LongName  \
2052   216325       A. Gunn       Angus Gunn   
6102   207804       A. Bell     Amari'i Bell   
17411  256151     D. Aquino    Derlis Aquino   
5200   238216  C. Gallagher  Conor Gallagher   

                                               photoUrl  \
2052   https://cdn.sofifa.com/players/216/325/21_60.png   
6102   https://cdn.sofifa.com/players/207/804/21_60.png   
17411  https://cdn.sofifa.com/players/256/151/21_60.png   
5200   https://cdn.sofifa.com/players/238/216/21_60.png   

                                               playerUrl Nationality  Age  \
2052   http://sofifa.com/player/216325/angus-gunn/210...     England   24   
6102   http://sofifa.com/player/207804/amarii-bell/21...     England   26   
17411  http://sofifa.com/player/256151/derlis-aquino/...    Paraguay   20   
5200   http://sofifa.com/player/238216/conor-gallaghe...     England   20   

       ↓OVA  POT                  Club              Contract Contract Start  \
2052     74   78           Southampton           2018 ~ 2023           2018   
6102     69   72      Blackburn Rovers           2018 ~ 2021           2018   
17411    56   66         Club Libertad           2019 ~ 2024           2019   
5200     70   83  West Bromwich Albion  Jun 30, 2021 On Loan            NaN   

      Contract End  Contract Length(years) Contract Status Positions Height  \
2052          2023                     5.0        Contract        GK  196cm   
6102          2021                     3.0        Contract        LB  180cm   
17411         2024                     5.0        Contract        CM  176cm   
5200           NaN                     0.0         On Loan   CAM, CM  182cm   

      Weight Preferred Foot  BOV Best Position     Joined Loan Date End  \
2052    77kg          Right   74            GK  10-Jul-18           NaN   
6102    77kg           Left   69            LB  22-Jan-18           NaN   
17411   74kg          Right   58           CAM  01-Jan-19           NaN   
5200    74kg          Right   72           CAM  07-Feb-17     30-Jun-21   

       Value  Wage Release Clause  Attacking  Crossing  Finishing  \
2052   €4.8M  €26K         €10.9M         94        14         15   
6102   €1.7M  €18K          €2.2M        271        64         47   
17411  €300K  €500          €336K        231        35         55   
5200   €3.8M  €29K             €0        273        49         66   

       Heading Accuracy  Short Passing  Volleys  Skill  Dribbling  Curve  \
2052                 14             32       19    123         20     20   
6102                 53             62       45    266         68     33   
17411                38             60       43    245         56     42   
5200                 48             70       40    322         73     52   

       FK Accuracy  Long Passing  Ball Control  Movement  Acceleration  \
2052            15            41            27       246            48   
6102            45            55            65       358            83   
17411           39            54            54       306            62   
5200            53            72            72       331            64   

       Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  Jumping  \
2052             49       40         67       42    231          53       64   
6102             77       63         64       71    318          55       74   
17411            62       66         56       60    256          44       49   
5200             62       67         70       68    320          62       57   

       Stamina  Strength  Long Shots  Mentality  Aggression  Interceptions  \
2052        31        66          17        140          28             22   
6102        76        64          49        273          59             64   
17411       53        53          57        237          39             55   
5200        75        60          66        329          68             67   

       Positioning  Vision  Pe

In [15]:
data[['Contract', 'Contract Start', 'Contract End', 'Contract Length(years)', 'Contract Status']].sample(5)

Contract Contract Start Contract End  \
8067   Jun 30, 2021 On Loan            NaN          NaN   
4603   Jun 30, 2021 On Loan            NaN          NaN   
12984           2016 ~ 2022           2016         2022   
12018           2020 ~ 2021           2020         2021   
4511   Jun 30, 2021 On Loan            NaN          NaN   

       Contract Length(years) Contract Status  
8067                      0.0         On Loan  
4603                      0.0         On Loan  
12984                     6.0        Contract  
12018                     1.0        Contract  
4511                      0.0         On Loan

### 3. Height

In [16]:
data['Height'].dtype

dtype('O')

In [17]:
data['Height'].unique()

array(['170cm', '187cm', '188cm', '181cm', '175cm', '184cm', '191cm',
       '178cm', '193cm', '185cm', '199cm', '173cm', '168cm', '176cm',
       '177cm', '183cm', '180cm', '189cm', '179cm', '195cm', '172cm',
       '182cm', '186cm', '192cm', '165cm', '194cm', '167cm', '196cm',
       '163cm', '190cm', '174cm', '169cm', '171cm', '197cm', '200cm',
       '166cm', '6\'2"', '164cm', '198cm', '6\'3"', '6\'5"', '5\'11"',
       '6\'4"', '6\'1"', '6\'0"', '5\'10"', '5\'9"', '5\'6"', '5\'7"',
       '5\'4"', '201cm', '158cm', '162cm', '161cm', '160cm', '203cm',
       '157cm', '156cm', '202cm', '159cm', '206cm', '155cm'], dtype=object)

Declare a function `transform_height` that accepts one parameter used to remove the string 'cm' and convert the height value from inches to centimeters, resulting integer data type only

In [18]:
def transform_height(x):
    if 'cm' in x:
        return int(x.strip('cm'))
    else:
        feet, inches = x.split("'")
        total_inches = int(feet)*12 + int(inches.strip('"'))
        return round(total_inches * 2.54)
    
data['Height'] = data['Height'].apply(transform_height)
data['Height'].unique()

array([170, 187, 188, 181, 175, 184, 191, 178, 193, 185, 199, 173, 168,
       176, 177, 183, 180, 189, 179, 195, 172, 182, 186, 192, 165, 194,
       167, 196, 163, 190, 174, 169, 171, 197, 200, 166, 164, 198, 201,
       158, 162, 161, 160, 203, 157, 156, 202, 159, 206, 155], dtype=int64)

In [19]:
data = data.rename(columns={'Height':'Height(Cm)'})
data.sample(1)

ID            Name                       LongName  \
1421  189156  Daniel Carriço  Daniel Filipe Martins Carriço   

                                              photoUrl  \
1421  https://cdn.sofifa.com/players/189/156/21_60.png   

                                              playerUrl Nationality  Age  \
1421  http://sofifa.com/player/189156/daniel-filipe-...    Portugal   31   

      ↓OVA  POT        Club     Contract Contract Start Contract End  \
1421    76   76  Wuhan Zall  2020 ~ 2021           2020         2021   

      Contract Length(years) Contract Status Positions  Height(Cm) Weight  \
1421                     1.0        Contract        CB         180   80kg   

     Preferred Foot  BOV Best Position     Joined Loan Date End Value  Wage  \
1421          Right   76            CB  20-Feb-20           NaN   €5M  €12K   

     Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
1421          €8.5M        304        61         48                73   

      Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  \
1421             75       47    290         58     39           54   

      Long Passing  Ball Control  Movement  Acceleration  Sprint Speed  \
1421            75            64       269            34            40   

      Agility  Reactions  Balance  Power  Shot Power  Jumping  Stamina  \
1421       58         71       66    330          66       84       53   

      Strength  Long Shots  Mentality  Aggression  Interceptions  Positioning  \
1421        77          50        342          85             76           57   

      Vision  Penalties  Composure  Defending  Marking  Standing Tackle  \
1421      62         62         73        236       79               78   

      Sliding Tackle  Goalkeeping  GK Diving  GK Handling  GK Kicking  \
1421              79           66         14           10          15   

      GK Positioning  GK Reflexes  Total Stats  Base Stats  W/F  SM     A/W  \
1421              15           12         1837         369  3 ★  2★  Medium   

         D/W   IR  PAC  SHO  PAS  DRI  DEF  PHY Hits  
1421  Medium  2 ★   37   53   67   61   78   73   24

### 4. Weight

In [20]:
data['Weight'].dtype

dtype('O')

In [21]:
data['Weight'].unique()

array(['72kg', '83kg', '87kg', '70kg', '68kg', '80kg', '71kg', '91kg',
       '73kg', '85kg', '92kg', '69kg', '84kg', '96kg', '81kg', '82kg',
       '75kg', '86kg', '89kg', '74kg', '76kg', '64kg', '78kg', '90kg',
       '66kg', '60kg', '94kg', '79kg', '67kg', '65kg', '59kg', '61kg',
       '93kg', '88kg', '97kg', '77kg', '62kg', '63kg', '95kg', '100kg',
       '58kg', '183lbs', '179lbs', '172lbs', '196lbs', '176lbs', '185lbs',
       '170lbs', '203lbs', '168lbs', '161lbs', '146lbs', '130lbs',
       '190lbs', '174lbs', '148lbs', '165lbs', '159lbs', '192lbs',
       '181lbs', '139lbs', '154lbs', '157lbs', '163lbs', '98kg', '103kg',
       '99kg', '102kg', '56kg', '101kg', '57kg', '55kg', '104kg', '107kg',
       '110kg', '53kg', '50kg', '54kg', '52kg'], dtype=object)

Declare a function `transform_weight` that removes the string 'kg' and 'lbs' characters, then converts the weight from pounds to kilograms, resulting integer data type

In [22]:
def transform_weight(x):
    if 'kg' in x:
        return int(x.strip('kg'))
    else:
        pounds = int(x.strip('lbs'))
        return round(pounds/2.205)
    
data['Weight'] = data['Weight'].apply(transform_weight)
data['Weight'].unique()

array([ 72,  83,  87,  70,  68,  80,  71,  91,  73,  85,  92,  69,  84,
        96,  81,  82,  75,  86,  89,  74,  76,  64,  78,  90,  66,  60,
        94,  79,  67,  65,  59,  61,  93,  88,  97,  77,  62,  63,  95,
       100,  58,  98, 103,  99, 102,  56, 101,  57,  55, 104, 107, 110,
        53,  50,  54,  52], dtype=int64)

In [23]:
data = data.rename(columns={'Weight':'Weight(kg)'})
data.sample(7)

ID        Name         LongName  \
13252  242482  E. Coronel  Emanuel Coronel   
14430  256407  F. Petroli   Franco Petroli   
16791  215104  A. Delgado   Álvaro Delgado   
15814  257153   A. Hausch   Agustin Hausch   
16258  255526   T. Kaneko    Takuro Kaneko   
3091   181262    M. Suchý      Marek Suchý   
16919  251520   I. Brînză     Ianoș Brînză   

                                               photoUrl  \
13252  https://cdn.sofifa.com/players/242/482/21_60.png   
14430  https://cdn.sofifa.com/players/256/407/21_60.png   
16791  https://cdn.sofifa.com/players/215/104/21_60.png   
15814  https://cdn.sofifa.com/players/257/153/21_60.png   
16258  https://cdn.sofifa.com/players/255/526/21_60.png   
3091   https://cdn.sofifa.com/players/181/262/21_60.png   
16919  https://cdn.sofifa.com/players/251/520/21_60.png   

                                               playerUrl     Nationality  Age  \
13252  http://sofifa.com/player/242482/emanuel-corone...       Argentina   23   
14430  http://sofifa.com/player/256407/franco-petroli...       Argentina   22   
16791  http://sofifa.com/player/215104/alvaro-delgado...           Chile   25   
15814  http://sofifa.com/player/257153/agustin-hausch...       Argentina   17   
16258  http://sofifa.com/player/255526/takuro-kaneko/...           Japan   22   
3091   http://sofifa.com/player/181262/marek-suchy/21...  Czech Republic   32   
16919  http://sofifa.com/player/251520/ianos-brinza/2...         Moldova   21   

       ↓OVA  POT                        Club              Contract  \
13252    62   70      Club Atlético Banfield           2019 ~ 2023   
14430    61   73                 River Plate           2020 ~ 2022   
16791    57   62              Audax Italiano           2020 ~ 2020   
15814    59   78      San Lorenzo de Almagro           2020 ~ 2022   
16258    58   65  Hokkaido Consadole Sapporo           2020 ~ 2021   
3091     72   72                 FC Augsburg           2019 ~ 2021   
16919    57   69            Politehnica Iaşi  Jun 30, 2021 On Loan   

      Contract Start Contract End  Contract Length(years) Contract Status  \
13252           2019         2023                     4.0        Contract   
14430           2020         2022                     2.0        Contract   
16791           2020         2020                     0.0        Contract   
15814           2020         2022                     2.0        Contract   
16258           2020         2021                     1.0        Contract   
3091            2019         2021                     2.0        Contract   
16919            NaN          NaN                     0.0         On Loan   

        Positions  Height(Cm)  Weight(kg) Preferred Foot  BOV Best Position  \
13252          RB         173          72          Right   64            CB   
14430          GK         186          84           Left   61            GK   
16791         CAM         175          64          Right   59           CAM   
15814      ST, LW         177          65          Right   61            ST   
16258  CF, RM, ST         178          68           Left   61           CAM   
3091           CB         183          76          Right   72            CB   
16919          GK         188          88          Right   57            GK   

          Joined Loan Date End  Value  Wage Release Clause  Attacking  \
13252  01-Jan-19           NaN  €850K   €2K          €680K        218   
14430  05-Oct-20           NaN  €700K   €2K          €560K         62   
16791  15-Jan-20           NaN  €240K   €1K          €200K        225   
15814  25-Jan-20           NaN  €575K  €700          €638K        262   
16258  01-Feb-20           NaN  €425K   €1K          €261K        240   
3091   08-Jul-19           NaN  €1.5M  €18K          €3.6M        275   
16919  25-Jul-17     30-Jun-21  €350K  €950             €0         64   

       Crossing  Finishing  Heading Accuracy  Short Passing  Volleys  Skill  \
13252        45         25                65 

### 5. Loan Date End

Date when the player's loan ends (if On Loan)

In [24]:
data['Loan Date End'].dtype

dtype('O')

In [25]:
data['Loan Date End'].unique()

array([nan, '30-Jun-21', '31-Dec-20', '30-Jan-21', '30-Jun-22',
       '31-May-21', '05-Jul-21', '31-Dec-21', '01-Jul-21', '01-Jan-21',
       '31-Aug-21', '31-Jan-21', '30-Dec-21', '23-Jun-21', '03-Jan-21',
       '27-Nov-21', '17-Jan-21', '30-Jun-23', '31-Jul-21', '22-Nov-20',
       '31-May-22', '30-Dec-20', '04-Jan-21', '30-Nov-20', '01-Aug-21'],
      dtype=object)

In [26]:
player_onloan = data[data['Contract Status'] == 'On Loan']
player_onloan[['Contract','Contract Status', 'Contract End', 'Loan Date End']]

Contract Contract Status Contract End Loan Date End
205    Jun 30, 2021 On Loan         On Loan          NaN     30-Jun-21
248    Jun 30, 2021 On Loan         On Loan          NaN     30-Jun-21
254    Jun 30, 2021 On Loan         On Loan          NaN     30-Jun-21
302    Jun 30, 2021 On Loan         On Loan          NaN     30-Jun-21
306    Jun 30, 2021 On Loan         On Loan          NaN     30-Jun-21
...                     ...             ...          ...           ...
18472  Aug 31, 2021 On Loan         On Loan          NaN     31-Aug-21
18571  Jun 30, 2021 On Loan         On Loan          NaN     30-Jun-21
18600  Dec 31, 2020 On Loan         On Loan          NaN     31-Dec-20
18622  Dec 31, 2020 On Loan         On Loan          NaN     31-Dec-20
18680  Dec 31, 2020 On Loan         On Loan          NaN     31-Dec-20

[1013 rows x 4 columns]

There are many missing values in the `Loan Date End` column because not all players in this dataset are players in the "On Loan" status. There are 1013 rows of data from players in the "On Loan" status, and I think this column might be useful in the future, so I won't delete it.

### 6. W/F

Player's weak foot rating (out of 5)

In [27]:
data['W/F'].dtype

dtype('O')

In [28]:
data['W/F'].unique()

array(['4 ★', '3 ★', '5 ★', '2 ★', '1 ★'], dtype=object)

In [29]:
data['W/F'] = data['W/F'].str.replace('★', '')
data['W/F'].unique()

array(['4 ', '3 ', '5 ', '2 ', '1 '], dtype=object)

### 7. SM

Player's skill moves rating (out of 5)

In [30]:
data['SM'].dtype

dtype('O')

In [31]:
data['SM'].unique()

array(['4★', '5★', '1★', '2★', '3★'], dtype=object)

In [32]:
data['SM'] = data['SM'].str.replace('★', '')
data['SM'].unique()

array(['4', '5', '1', '2', '3'], dtype=object)

### 8. Hits

Number of times the player has been searched for in the FIFA database.

In [33]:
data['Hits'].dtype

dtype('O')

In [34]:
data['Hits'].unique()

array(['771', '562', '150', '207', '595', '248', '246', '120', '1.6K',
       '130', '321', '189', '175', '96', '118', '216', '212', '154',
       '205', '202', '339', '408', '103', '332', '86', '173', '161',
       '396', '1.1K', '433', '242', '206', '177', '1.5K', '198', '459',
       '117', '119', '209', '84', '187', '165', '203', '65', '336', '126',
       '313', '124', '145', '538', '182', '101', '45', '377', '99', '194',
       '403', '414', '593', '374', '245', '3.2K', '266', '299', '309',
       '215', '265', '211', '112', '337', '70', '159', '688', '116', '63',
       '144', '123', '71', '224', '113', '168', '61', '89', '137', '278',
       '75', '148', '176', '197', '264', '214', '247', '402', '440',
       '1.7K', '2.3K', '171', '320', '657', '87', '259', '200', '255',
       '253', '196', '60', '97', '85', '169', '256', '132', '239', '166',
       '121', '109', '32', '46', '122', '48', '527', '199', '282', '51',
       '1.9K', '642', '155', '323', '288', '497', '509', '79',

Declare a function `convert_hits` that replaces null values with 0, removes the string 'k' character, multiplies it by 1000, and converts it to an integer data type. Finally, if there's a float or integer value, it will be converted to an integer value only.

In [35]:
def convert_hits(x):
    if pd.isnull(x):
        return 0
    elif 'K' in str(x):
        number_only = str(x).replace('K', '')
        number_float = float(number_only)
        number_int = int(number_float * 1000)
        return number_int
    elif isinstance(x, float) or isinstance(x, int):
        return int(x)
    else:
        return int(float(x))

    
data['Hits'] = data['Hits'].apply(convert_hits)


In [36]:
data['Hits'].dtype

dtype('int64')

In [37]:
data['Hits'].unique()

array([ 771,  562,  150,  207,  595,  248,  246,  120, 1600,  130,  321,
        189,  175,   96,  118,  216,  212,  154,  205,  202,  339,  408,
        103,  332,   86,  173,  161,  396, 1100,  433,  242,  206,  177,
       1500,  198,  459,  117,  119,  209,   84,  187,  165,  203,   65,
        336,  126,  313,  124,  145,  538,  182,  101,   45,  377,   99,
        194,  403,  414,  593,  374,  245, 3200,  266,  299,  309,  215,
        265,  211,  112,  337,   70,  159,  688,  116,   63,  144,  123,
         71,  224,  113,  168,   61,   89,  137,  278,   75,  148,  176,
        197,  264,  214,  247,  402,  440, 1700, 2300,  171,  320,  657,
         87,  259,  200,  255,  253,  196,   60,   97,   85,  169,  256,
        132,  239,  166,  121,  109,   32,   46,  122,   48,  527,  199,
        282,   51, 1900,  642,  155,  323,  288,  497,  509,   79,   49,
        270,  511,   80,  128,  115,  156,  204,  143,  140,  152,  220,
        134,  225,   94,   74,  135,  142,   50,   

### Look of all columns

In [38]:
data.sample(5)

ID          Name                 LongName  \
14138  255038   H. Nakamura          Hotaka Nakamura   
13333  176472    N. Cabrera           Nelson Cabrera   
4142   211060       Elsinho  Elson Ferreira de Souza   
1149   206225  D. Cheryshev          Denis Cheryshev   
8429   234751   N. Breeveld        Nicandro Breeveld   

                                               photoUrl  \
14138  https://cdn.sofifa.com/players/255/038/21_60.png   
13333  https://cdn.sofifa.com/players/176/472/21_60.png   
4142   https://cdn.sofifa.com/players/211/060/21_60.png   
1149   https://cdn.sofifa.com/players/206/225/21_60.png   
8429   https://cdn.sofifa.com/players/234/751/21_60.png   

                                               playerUrl  Nationality  Age  \
14138  http://sofifa.com/player/255038/hotaka-nakamur...        Japan   22   
13333  http://sofifa.com/player/176472/nelson-cabrera...      Bolivia   37   
4142   http://sofifa.com/player/211060/elson-ferreira...       Brazil   30   
1149   http://sofifa.com/player/206225/denis-cheryshe...       Russia   29   
8429   http://sofifa.com/player/234751/nicandro-breev...  Netherlands   33   

       ↓OVA  POT              Club     Contract Contract Start Contract End  \
14138    61   66          FC Tokyo  2020 ~ 2021           2020         2021   
13333    62   62      Always Ready  2019 ~ 2023           2019         2023   
4142     71   71   Shimizu S-Pulse  2019 ~ 2022           2019         2022   
1149     77   77       Valencia CF  2019 ~ 2022           2019         2022   
8429     67   67  Politehnica Iaşi  2019 ~ 2022           2019         2022   

       Contract Length(years) Contract Status Positions  Height(Cm)  \
14138                     1.0        Contract    RB, LB         177   
13333                     4.0        Contract    CB, RB         191   
4142                      3.0        Contract    RB, RM         180   
1149                      3.0        Contract        LM         179   
8429                      3.0        Contract       CDM         178   

       Weight(kg) Preferred Foot  BOV Best Position     Joined Loan Date End  \
14138          72          Right   63            CB  01-Feb-20           NaN   
13333          84           Left   62            CB  01-Jan-19           NaN   
4142           78          Right   71            RB  10-Jan-19           NaN   
1149           74           Left   77            LM  12-Jul-19           NaN   
8429           76          Right   67           CDM  11-Jul-19           NaN   

        Value  Wage Release Clause  Attacking  Crossing  Finishing  \
14138   €475K   €2K          €398K        224        59         22   
13333    €70K  €500          €129K        213        32         32   
4142    €1.6M   €3K          €2.4M        307        73         58   
1149   €10.5M  €35K         €17.9M        357        84         72   
8429    €500K   €3K          €506K        279        53         46   

       Heading Accuracy  Short Passing  Volleys  Skill  Dribbling  Curve  \
14138                59             54       30    198         51     27   
13333                65             61       23    204         37     31   
4142                 57             68       51    339         72     59   
1149                 58             74       69    365         74     76   
8429                 58             65       57    272         54     44   

       FK Accuracy  Long Passing  Ball Control  Movement  Acceleration  \
14138           27            44            49       310            64   
13333           40            49            47       265            60   
4142            63            71            74       331            69   
1149            61            77            77       381            75   
8429            52            65            57       336            67   

       Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  Jumping  \
14138            66       58         55       67    265         